File: ArchitectureComponentsMVVM/app/src/main/java/com/andr/mvvm/repository/UserRoomDataBase.java
package com.andr.mvvm.repository;

import android.arch.persistence.db.SupportSQLiteDatabase;
import android.arch.persistence.room.Database;
import android.arch.persistence.room.Room;
import android.arch.persistence.room.RoomDatabase;
import android.content.Context;
import android.support.annotation.NonNull;

@Database(entities = {User.class}, version = 1)
public abstract class UserRoomDataBase extends RoomDatabase {
    public abstract UserDao userDao();

    private static UserRoomDataBase INSTANCE;


    static UserRoomDataBase getDatabase(final Context context) {
        if (INSTANCE == null) {
            synchronized (UserRoomDataBase.class) {
                if (INSTANCE == null) {
                    INSTANCE = Room.databaseBuilder(context.getApplicationContext(),
                            UserRoomDataBase.class, "user_database")
                            .addCallback(sRoomDatabaseCallback)
                            .build();

                }
            }
        }
        return INSTANCE;
    }

    private static RoomDatabase.Callback sRoomDatabaseCallback =
            new RoomDatabase.Callback(){

                @Override
                public void onOpen (@NonNull SupportSQLiteDatabase db){
                    super.onOpen(db);
                    new PopulateDbAsync(INSTANCE).execute();
                }
            };
}

File: MVVM_Kotlin/app/src/main/java/com/mvvm/kot/Repository/UserRoomDataBase.kt
package com.mvvm.kot.Repository

import android.arch.persistence.db.SupportSQLiteDatabase
import android.arch.persistence.room.Database
import android.arch.persistence.room.Room
import android.arch.persistence.room.RoomDatabase
import android.content.Context
import com.mvvm.kot.ViewModels.User

@Database(entities = arrayOf(User::class), version = 1)
abstract  class UserRoomDataBase: RoomDatabase(){
    abstract fun userDao(): UserDao

    companion object {
        private var INSTANCE: UserRoomDataBase? = null

        fun getInstance(context: Context): UserRoomDataBase? {
            if (INSTANCE == null) {
                synchronized(UserRoomDataBase::class) {
                    INSTANCE = Room.databaseBuilder(context,
                            UserRoomDataBase::class.java, "userData.db")
                            .addCallback(sRoomDataBaseCallback)
                            .build()
                }
            }
            return INSTANCE
        }

        fun destroyInstance() {
            INSTANCE = null
        }

        val sRoomDataBaseCallback = object:RoomDatabase.Callback(){
            override fun onOpen(db: SupportSQLiteDatabase) {
                super.onOpen(db)
                InitDBAsync(INSTANCE!!).execute()
            }
        };
     /* val sRoomDatabaseCallback = object : RoomDatabase.Callback() {

            override fun onOpen(db: SupportSQLiteDatabase) {
                super.onOpen(db)
                InitDBAsync(INSTANCE).execute()
            }
        }*/
    }

    }
